### PR TITLE
Image editing: batch editing in cropper component

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

The difference with #22959 is that this uses the cropper for all edits. To do this, I've removed flipping for now, which is probably a lesser used feature anyway. I copied the PHP from the other PR. The purpose of this PR is to unblock iterations and possibly remove the experimental flag for the next GB release.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
Would we then be able to drop the specific edit endpoints and just use `/apply`?

**Comment by ajlende** [comment]
When moving all edits within react-easy-crop there isn't an option to rotate without cropping to one of the preset aspect ratios. I think that you're right about the flip probably being a lesser used feature, but I think it's important to be able to rotate without cropping.

File: lib/class-wp-rest-image-editor-controller.php [hunk 1]
```diff
@@ -116,6 +116,72 @@ public function register_routes() {
...
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/apply',
+			array(
+				array(
+					'methods'             => WP_REST_Server::EDITABLE,
+					'callback'            => array( $this, 'apply_edits' ),
+					'permission_callback' => array( $this, 'permission_callback' ),
+					'args'                => array(
+						array(
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:129]: This API is confusing, but when specifying `args` it already assumes that the overall type is an object, and the `args` array is a collection of property names => schemas.

So we should drop the overall `[ type => 'object', 'properties' => [] ]` and instead have it be `args.crop => []` etc...

Comment (ajlende) [reply lib/class-wp-rest-image-editor-controller.php:129]: Does doing that make the property required?

Comment (TimothyBJacobs) [reply lib/class-wp-rest-image-editor-controller.php:129]: No, properties aren't required by default, only if `required` is set for that property name.

Not required
```php
$args = [
  'crop' => [
      'type' => 'object',
     'properties' => [],
   ]
];
```

Required
```php
$args = [
  'crop' => [
      'type' => 'object',
      'required' => true,
      'properties' => [],
   ]
];
```

**Comment by ellatrix** [comment]
@ajlende I’ll add it tomorrow morning

**Comment by ellatrix** [comment]
@TimothyBJacobs yes let’s drop the other endpoints and also keep the whole api experimental

**Comment by mtias** [comment]
I like this path, it allows us to simplify the flow a bit (going through the _Apply_ action). I also think it's good to mark the endpoints experimental, it seems like a model we should be willing to try in general as it'll allow us to gather feedback from developers widely during a cycle before committing to a final shape.

**Comment by mtias** [comment]
A few notes:

- This button still shows up with the experiment disabled.

image

image

- The aspect ratio is not working very well for me (gets reverted after "apply").
- Can we improve the increments of the "zoom"? The jumps between steps are too abrupt and the number is not very intuitive (percentage would be better, I think).
- It should likely use @ItsJonQ input number control set to percentage (so it shows the percentage sign).

**Design:**

image

- I'd still use the "crop" icon instead of "edit".
- We should move "Replace" to the right, after "link" and after "crop". Seems awkward there.
- Cancel and Apply should be grouped in the toolbar, and in the right order.
- We should hide all the other toolbar options when entering into edit mode (alignment, ellipsis menu)
- The zoom component should be rendered in the toolbar.

**Comment by ellatrix** [comment]
@mtias Yes, I see some issues too when applying the image. Looks like something isn't processed on the server. I'll have a look.

Agree on Zoom. Let's try to move it to the toolbar in a separate PR.

**Comment by ellatrix** [comment]
> This button still shows up with the experiment disabled.

Fixed

> The aspect ratio is not working very well for me (gets reverted after "apply").

Not sure if I have the same problem, but I do have problems with square aspect ratio, which I'm looking into now. Seems like sometimes cropping on the server is not correct, probably instructions being processed wrong.

> Can we improve the increments of the "zoom"? The jumps between steps are too abrupt and the number is not very intuitive (percentage would be better, I think).
> It should likely use @ItsJonQ input number control set to percentage (so it shows the percentage sign).

Let's do that when we work on the other zoom things.

> I'd still use the "crop" icon instead of "edit".

Fixed

> We should move "Replace" to the right, after "link" and after "crop". Seems awkward there.

Fixed

> Cancel and Apply should be grouped in the toolbar, and in the right order.

Fixed

> We should hide all the other toolbar options when entering into edit mode (alignment, ellipsis menu)

Partly fixed. The general block buttons will have to be hidden in a special way as they're out of the block's control. Removed the other buttons though.

> The zoom component should be rendered in the toolbar.

Separate PR

**Comment by ellatrix** [comment]
There's also issues with cropping in current `master`. When I crop to a different aspect ratio, it turns out much longer. Not sure why. There's some issues on the server side.

image-miscrop

**Comment by TimothyBJacobs** [comment]
So is this creating a new record in the media library?

**Comment by ellatrix** [comment]
@TimothyBJacobs Once you apply the edits, yes

**Comment by TimothyBJacobs** [comment]
In that case, can we instead return the results of a `rest_do_request( '/wp/v2/media/{id}' )` and add a `Location` header with a `201` status code instead of the array? Something like.

```php
$path = '/wp/v2/media/' . $attachment_id;
$response = rest_do_request( $path );

if ( ! $response->is_error() ) {
    $response->set_status( 201 );
    $response->header( 'Location', rest_url( $path ) ); 
}
```

**Comment by ellatrix** [comment]
@TimothyBJacobs Let's review the REST API endpoint separately. For this PR, I'd like to focus on doing the edits in a single request on apply and some user facing polish. I'm going to merge this PR since we'd like to include it for the Gutenberg release.

**Comment by TimothyBJacobs** [comment]
Sure thing, is there an open issue for the PHP side of things? Does this invalidate #22959? I'm having trouble tracking the related issues.

> I'm going to merge this PR since we'd like to include it for the Gutenberg release.

The endpoint's return value would impact the client code. So if this Gutenberg release is the same as the 5.5 release, it'll need to be addressed before then as well.

**Pull request merged**